### PR TITLE
[RFC] First small steps to arbitrary git repositories

### PR DESCRIPTION
**Pull request by bstoeger**
### Describe the pull request:

- [ ] Bug fix
- [x] Functional change
- [ ] New feature
- [ ] Code cleanup
- [ ] Build system change
- [ ] Documentation change
- [ ] Language translation

### Pull request long description:

These are first steps towards supporting arbitrary git repositories, here especially ssh:// repositories. This also unifies username handling of ssh:// and https:// repositories.

A major change is the handling of authentication. The code used to only support private-key based authentication. Now, if the private key does not exist, the cloud password is used. In the future, this should of course be handled on a per-repository basis.

The PR as-is is not yet useful without UI changes. With a different set of commits, I managed to save to a remote ssh:// repository. Though I had to check out a dummy branch - saving to checked-out master failed without error message(!). So it's all still a bit brittle.

I'm posting this for comments. I would be especially grateful if people using the private-key based ssh:// transport (if this actually used to work) could test the changes.

### CODE REVIEW

**Comment by mturkia** [comment]
How do I set up such and SSH repo configuration in Subsurface?

**Comment by bstoeger** [comment]
I'm accessing a ssh://-style repository using a different patch-set, which is not yet ready. Among other things there are still two major problems:
1) All git repositories use the same password (the cloud password from preferences).
2) It breaks mobile.

If it would be OK to only support username/password style authentication at first, that would make things easier.

> And ... "tested" this PR for regressions, and it seems just fine in that respect.

Thanks. It shouldn't make any difference for cloud access, but good to have that confirmed.

File: core/git-access.c [hunk 1]
```diff
@@ -195,19 +195,43 @@ int credential_ssh_cb(git_cred **out,
...
-		return GIT_EUSER;
+	const char *username = prefs.cloud_storage_email_encoded;
+	const char *passphrase = prefs.cloud_storage_password ? prefs.cloud_storage_password : "";
+
+	// TODO: We need a way to differentiate between password and private key authentication
+	if (allowed_types & GIT_CREDTYPE_SSH_KEY) {
+		char *priv_key = format_string("%s/%s", system_default_directory(), "ssrf_remote.key");
+		if (!access(priv_key, F_OK)) {
+			/* Bail out from libgit authentication loop when credentials are incorrect */
+			if (auth_attempt++ > 2) {
```
Comment (neolit123) [inline core/git-access.c]: maybe we should hava a re-usable macro or a constant for the `2` (max number of attempts) ?

File: core/git-access.c [hunk 2]
```diff
@@ -195,19 +195,43 @@ int credential_ssh_cb(git_cred **out,
...
+		if (auth_attempt++ > 2) {
+			report_error("Authentication to cloud storage failed.");
+			return GIT_EUSER;
+		}
+		return git_cred_userpass_plaintext_new(out, username, passphrase);
 	}
 
-	const char *priv_key = format_string("%s/%s", system_default_directory(), "ssrf_remote.key");
-	const char *passphrase = prefs.cloud_storage_password ? strdup(prefs.cloud_storage_password) : strdup("");
+	if (allowed_types & GIT_CREDTYPE_USERNAME) {
```
Comment (neolit123) [inline core/git-access.c]: no need for these `{ }`.

File: core/git-access.c [hunk 3]
```diff
@@ -794,18 +821,16 @@ static struct git_repository *get_remote_repo(const char *localdir, const char *
...
-			prefs.git_local_only = glo;
-			return ret;
-		}
+		/* We have no local cache yet.
+		 * Take us temporarly online to create a local and
+		 * remote cloud repo.
+		 */
+		git_repository *ret;
+		bool glo = prefs.git_local_only;
+		prefs.git_local_only = false;
```
Comment (neolit123) [inline core/git-access.c:824]: maybe `const bool glo` but that's just a nitpick and it may produce a warning.

**Review by neolit123 — COMMENTED** [review]
added a few small style comments.

File: core/git-access.c (hunk 1 above)
Comment (bstoeger) [reply core/git-access.c]: Factored this out into a function and made `max_auth_attempts` a static variable. For C++ I'm sure that this would be compiled out, so no different from a macro - for C I'm not sure, but will check.

File: core/git-access.c (hunk 2 above)
Comment (bstoeger) [reply core/git-access.c]: Removed.

File: core/git-access.c (hunk 3 above)
Comment (bstoeger) [reply core/git-access.c:824]: I'll leave this change to the original author (@dirkhh, I presume). The problem I see is that, if we start `const`ifying local variables, this will take no end...

File: core/git-access.c (hunk 1 above)
Comment (neolit123) [reply core/git-access.c]: gcc will probably preserve the branch (function) without optimization flags.

File: core/git-access.c (hunk 3 above)
Comment (neolit123) [reply core/git-access.c:824]: understood.

File: core/git-access.c [hunk 4]
```diff
@@ -187,6 +187,16 @@ static int reset_to_remote(git_repository *repo, git_reference *local, const git
 }
 
 static int auth_attempt = 0;
+static const int max_auth_attempts = 2;
+
+static bool exceeded_auth_attempts()
+{
```
Comment (neolit123) [inline core/git-access.c:193]: i think this is fine.

**Comment by janmulder** [comment]
> If it would be OK to only support username/password style authentication at first, that would make things easier.

A small reaction on this (without testing the latest push to this PR). We can assume that users who want to use an own (private) git server for the cloud storage can setup a simple ssh connection with passwd or with a key pair outside of Subsurface. In other words, we do not need a fancy UI for this just to start testing this. We basically only need an extra toggle in the preferences on the general tab (use own git server), and the capability to specify the ssh url, or an own https capable git repo (whatever the user wants).

So basically, I do not really understand the fundamental complexity here.

Basically, no other developers will step in when there is no easy way to wire an own git server from the UI, or  even a stupid manual hack in some file, as, like I said, it is not to expect that there are a lot ssh and/or own git server users (as I believe that any support for this in the current master is well hidden and sparsely documented, even for developers).

**Comment by janmulder** [comment]
I hacked around on top of this PR, and succeeded in using my own git sever over ssh, both key based and password based. Working for both mobile-on-desktop and desktop, but a true hurdle will be mobile-on-device. Perfectly happy to push this WIP to github, but definitely not to make a PR from it (as it is @bstoeger PR and he also has WIP on top of this, as can been seen above).

This said, we might consider this PR as finished and ready to merge, as it seems stable with respect to changes, and it does not break things, but fixes things very useful for further use instead. I more and more believe that nobody is using the ssh git stuff that is currently sitting in master, as I 'm pretty sure that that is broken as it is now.

**Comment by bstoeger** [comment]
Thank you so much - I didn't find time to backport my UI stuff since I was caught up in work & private life and my code is still very much in flux.

**Comment by janmulder** [comment]
My stuff is sitting here: https://github.com/janmulder/subsurface/commits/pr/1044%2BJLMcommits

The idea is:

- introduce a new default_file_state, UI management of this toggle is to be done, but trivial. But to use above code, you currently have to edit the preference file by hand (as the is no UI yet) to set the correct new state.
- use the current "Local dive log file" as the URL input for any repo
- currently supported is the format: `ssh://<user@host><path/to/repo>[branch]`
- for key login: user has to place the key in the proper place, on desktop this will stay this way, why clutter (any) UI to manage the private key on client side? So, as already said: when the user does not want key login any more, user has to remove the key.
- for passwd login: mis-use the passwd from the "network" tab to ssh push/fetch from the git server.

Main hurdle is device-on-device, and noting is done by me (yet) on this. Even when using the passwd from the official cloud account, we still need to be able to fill in the (long) URL to the private git repo. There is no support for that yet. Shouldn't be too complex, but that mobile code is tricky to say the least. For example, for own git server, there is no PIN phase, and it needs not to interfere with the no-cloud setup. 

And (in case of key login) as a normal user has no access to any of the files from the app, we need to generate a key pair from the app, present the public key to the user who can put that on his git server.

**Comment by dirkhh** [comment]
Sorry, I have been distracted with other things.
A few small comments that may make you sad... I have been building several of the binaries (including Android and Mac) without ssh support lately because no one used it. And because this caused quite a bit of pain in the dependencies (ssh support had been broken for more than a year on Mac for example - and of course no one used it and no one noticed it).
I think that as a hack for the six people who want this that's not a bad thing, but as something to explain to regular users? Heck no. Please don't go there.
So in my mind this has to be a compile time option that will be OFF for the binaries that I distribute and people who really want to set up an ssh-authenticated git server because they can not fathom the thought that I might be able to see their dives will have to please build from sources.

I'm of course open to comments on this, but I'll admit that I feel pretty strongly about this.

**Comment by bstoeger** [comment]
> A few small comments that may make you sad

Not really. I don't see the problem with putting the ssh-specific code in `#ifdefs`. But I will do this when everything else is finished (not very soon, since I lack time myself). Request for general git repositories comes from @atdotde: https://github.com/Subsurface-divelog/subsurface/pull/928#issuecomment-350539088 and it's not particularly difficult to implement, so why not? Note that the driving force is not paranoia.

The particular PR here does not have user-visible changes anyway, since it only concerns non-https git repositories.

**Comment by glance-** [comment]
I'm fine with having ifdef's around the SSH-code but wouldn't it be better if we could leave it default, when we've fixed the current issues?

If we leave it default off, the code will rot, and the author which breaks it won't even be aware of it.

The upside with having it ifdef'ed is that its easy to just disable it for a release build, when / if it ends up causing us problems.

**Comment by janmulder** [comment]
I believe this PR (not the stuff I have sitting on the mentioned branch of mine) is formally unrelated to ssh. It fixes broken code that is not ifdef-ed at all (in current master). Please read the commits. As Berthold says, no visible UI changes. That ssh it not going to work on Android or Apple as ssh things are not compiled in, so be it, that is another discussion. This PR does not change anything in this respect.

I further agree with @dirkhh that not exposing all kinds Git transport options explicitly to the normal user is fine. But there seems no reason to force the other Git transport users (and also agree with Dirk that there very few of that type of uses) towards self building and introducing a new define to Subsurface to disable specifically ssh to make their life even harder.

As a side note. Why is ssh transport not working on Android? I have a perfectly working ssh client app on my Android phone.

**Comment by atdotde** [comment]
I should add that the purpose that I have in mind (small dongle to do the non-BT downloading for mobile) doing it over https rather than ssh is not a show stopper (as we would control the set-up of the other side). But all this is very speculative as I do not even have a working proof of concept laking the time.

**Comment by bstoeger** [comment]
Yes, but note that only the last commit in the series specifically concerns ssh repositories:
- Commit 1 and 2 concern https://-style non-cloud repositories
- Commit 3 concerns any non-https://-style repository (including e.g. http:// or git://)
- Commit 4 is a general cleanup
- Commit 5 is mostly ssh:// specific (introduces username/password style authentication), but shouldn't break any working usecase.
 
I shouldn't have mentioned ssh:// so prominently in the PR description. :(

**Comment by dirkhh** [comment]
I think this is a good PR. The last commit worries me a tiny bit, but it doesn't mess up the git storage test, so hopefully it's good.
